cleverhans/experimental/certification/dual_formulation.py

```diff
@@ -282,11 +288,13 @@ def get_psd_product(self, vector):
       result_product: Matrix product of M and vector
     """
     # For convenience, think of x as [\alpha, \beta]
+    vector = tf.cast(vector, self.nn_dtype)
     alpha = tf.reshape(vector[0], shape=[1, 1])
     beta = vector[1:]
     # Computing the product of matrix_h with beta part of vector
     # At first layer, h is simply diagonal
     h_beta = self.get_h_product(beta)
+    h_beta = tf.cast(h_beta, self.nn_dtype)
```

Comment:
Is this line needed? 
`self.get_h_produce(beta)` already should cast everything to `nn_dtype`

Also you can do following:
`h_beta = self.get_h_product(beta, dtype=self.nn_dtype)`